crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct ValidatedSearchPathSettings {
+    extra_paths: Vec<SearchPathRoot>,
+    workspace_root: SearchPathRoot,
+    stdlib: SearchPathRoot,
+    site_packages: Option<SearchPathRoot>,
+}
+
+impl ValidatedSearchPathSettings {
+    /// Implementation of the typing spec's [module resolution order].
+    ///
+    /// It's important that we collect these instead of iterating over them lazily,
+    /// so that we can be sure that it doesn't contain any duplicates.
+    /// At runtime, [`sys.path` never contains duplicate entries]
+    ///
+    /// [module resolution order]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
+    /// [`sys.path` never contains duplicate entries]: https://docs.python.org/3/library/site.html#module-site
+    fn static_search_paths(&self) -> OrderSet<&SearchPathRoot> {
+        let ValidatedSearchPathSettings {
+            extra_paths,
+            workspace_root,
+            stdlib,
+            site_packages,
+        } = self;
+        extra_paths
+            .iter()
+            .chain(std::iter::once(workspace_root))
+            .chain(std::iter::once(stdlib))
+            .chain(site_packages)
+            .collect()
+    }
+}
+
+/// Collect all dynamic search paths:
+/// search paths listed in `.pth` files in the `site-packages` directory
+/// due to editable installations of third-party packages.
+#[salsa::tracked(return_ref)]
+pub(crate) fn dynamic_module_resolution_paths(db: &dyn Db) -> OrderSet<SearchPathRoot> {
+    let settings = &module_resolver_settings(db).search_path_settings;
+    let static_paths = settings.static_search_paths();
+    let mut dynamic_paths = OrderSet::default();
+
+    if let Some(site_packages) = &settings.site_packages {
+        let site_packages = site_packages
+            .as_system_path()
+            .expect("Expected site-packages never to be a VendoredPath!");
+
+        // As well as modules installed directly into `site-packages`,
+        // the directory may also contain `.pth` files.
+        // Each `.pth` file in `site-packages` may contain one or more lines
+        // containing a (relative or absolute) path.
+        // Each of these paths may point to an editable install of a package,
+        // so should be considered an additional search path.
+        let Ok(pth_file_iterator) = PthFileIterator::new(db, site_packages) else {
+            return dynamic_paths;
+        };
+
+        // The Python documentation specifies that `.pth` files in `site-packages`
+        // are processed in alphabetical order.
+        // We know that files returned from [`ruff_db::system::MemoryFileSystem::read_directory()`]
+        // will be in alphabetical order already. However, [`ruff_db::system::OsSystem::read_directory()`]
+        // gives no such guarantees. Therefore, collecting and then sorting is necessary.
+        // https://docs.python.org/3/library/site.html#module-site
```

Comment:
The `MemoryFileSystem` also doesn't guarantee this in its contract. It would be okay for it to shuffle the entries before returning. 

I recommend shortening the comment. The first two lines are what's important.